src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1892,6 +1913,99 @@ private boolean unsignedLongCompare(long one, long two) {
+            final double base = Math.nextUp(x >= 0 ? x : x + 0x1p64);
+            final double exp = Math.nextUp(1.0 / n);
+            long rLong = (long) Math.ceil(Math.nextUp(Math.pow(base, exp)));
+
+            if (BigInteger.bitLengthForLong(rLong) * n <= Long.SIZE) {
+                // Refine the estimate.
+                do {
+                    long rToN1 = BigInteger.unsignedLongPow(rLong, n - 1);
+                    long rToN = rToN1 * rLong;
+                    if (Long.compareUnsigned(rToN, x) <= 0)
+                        return new MutableBigInteger(rLong);
+
+                    // compute rLong - ceil((rToN - x) / (n * rToN1))
+                    long dividend = rToN - x, divisor = n * rToN1;
+                    if (Long.remainderUnsigned(dividend, divisor) != 0)
+                        rLong--;
+
+                    rLong -= Long.divideUnsigned(dividend, divisor);
+                } while (true);
+            } else { // r^n could overflow long range, use MutableBigInteger loop instead
+                r = new MutableBigInteger(rLong);
+            }
+        } else {
+            // Set up the initial estimate of the iteration.
+            // Determine a right shift that is a multiple of n into finite double range.
+            long shift = Math.max(0, bitLength - Double.MAX_EXPONENT); // use long to avoid overflow later
+            int shiftExcess = (int) (shift % n);
+
+            // Shift the value into finite double range
+            double base = this.toBigInteger().shiftRight((int) shift).doubleValue();
+            // Complete the shift to a multiple of n,
+            // avoiding to lose more bits than necessary.
+            if (shiftExcess != 0) {
+                int shiftLack = n - shiftExcess;
+                shift += shiftLack; // shift is long, no overflow
+                base /= Double.valueOf("0x1p" + shiftLack);
+            }
+
+            // Use the root of the shifted value as an estimate.
+            base = Math.nextUp(base);
+            final double exp = Math.nextUp(1.0 / n);
+            r = valueOf(Math.ceil(Math.nextUp(Math.pow(base, exp))));
+
+            // Shift the approximate root back into the original range.
+            r.safeLeftShift((int) (shift / n));
+        }
+
+        // Refine the estimate.
+        do {
+            BigInteger rBig = r.toBigInteger();
+            BigInteger rToN1 = rBig.pow(n - 1);
+            MutableBigInteger rToN = new MutableBigInteger(rToN1.multiply(rBig).mag);
+            if (rToN.subtract(this) <= 0)
+                return r;
+
+            // compute r - ceil((r^n - this) / (n * r^(n - 1)))
+            MutableBigInteger q1 = new MutableBigInteger();
+            MutableBigInteger delta = new MutableBigInteger();
+            // Don't use conditional-or to ensure to do both divisions
+            if (rToN.divideOneWord(n, q1) != 0
+                    | !q1.divide(new MutableBigInteger(rToN1.mag), delta).isZero())
```

Comment:
```suggestion
            // Don't use conditional-or to ensure to do both divisions [because ...]
            if ((rToN.divideOneWord(n, q1) != 0)
                    | (!q1.divide(new MutableBigInteger(rToN1.mag), delta).isZero()))
```
... add parens for clarity.